src/graph/commands/barchart/barchart.pas

```diff
@@ -0,0 +1,220 @@
+unit barchart;
+
+{$mode objfpc}{$H+}
+
+interface
+
+uses
+  Classes, SysUtils, chartcommandresult, TAGraph, executor, outputcreator,
+  ast, epidatafiles, chartcommand, chartfactory, chartconfiguration,
+  tables_types, tables;
+
+type
+
+  { TBarChart }
+
+  TBarChart = class(TInterfacedObject, IChartCommand)
+  private
+    FChartFactory: IChartFactory;
+    FExecutor: TExecutor;
+    FOutputCreator: TOutputCreator;
+  public
+    procedure Init(ChartFactory: IChartFactory; Executor: TExecutor; OutputCreator: TOutputCreator);
+    function Execute(Command: TCustomGraphCommand): IChartCommandResult;
+  end;
+
+implementation
+
+uses
+  TASeries, TASources, TATypes, TAStyles, TAChartUtils, TALegend,
+  Graphics, charttitles, ast_types, epidatafilestypes,
+  epicustombase, epifields_helper, options_utils, graph_utils, barchartsource;
+
+{ TBarChart }
+
+procedure TBarChart.Init(ChartFactory: IChartFactory; Executor: TExecutor;
+  OutputCreator: TOutputCreator);
+begin
+  FChartFactory := ChartFactory;
+  FExecutor := Executor;
+  FOutputCreator := OutputCreator;
+end;
+
+function TBarChart.Execute(Command: TCustomGraphCommand): IChartCommandResult;
+const
+  dummyVarName = '_dummy4barchart';
+var
+  {Chart}
+  Chart:               TChart;
+  ChartConfiguration:  IChartConfiguration;
+  BarSource:           TBarSource;
+  LabelSeries:         TListChartSource;
+  BarSeries:           TBarSeries;
+  SeriesStyles:        TChartStyles;
+  aStyle:              TChartStyle;
+  sColor:              TColorMap;
+  {Frequencies}
+  DataFile:            TEpiDataFile;
+  T:                   TTables;
+  StratVariable:       TStringList;
+  nilTablesRefMap:     TEpiReferenceMap;
+  TablesAll:           TTwoWayTables;
+  TableData:           TTwoWayTable;
+  {command}
+  VarNames:            TStrings;
+  DFVars:              TStrings;
+  XVar:                TEpiField;
+  dummyVar:            TEpiField;
+  WeightVarName:       UTF8String;
+  cOptions:            TOptionList;
+  Opt:                 TOption;
+
+  ValueLabelOutput:    TEpiGetValueLabelType;
+  VariableLabelOutput: TEpiGetVariableLabelType;
+  ReverseStrata:       Boolean;
+  ByVarName:           UTF8String;
+  i, colourNum:           Integer;
+  sTitle:              UTF8String;
+  yPct:                Boolean;
+  yType:               UTF8String;
+begin
+  VariableLabelOutput := VariableLabelTypeFromOptionList(Command.Options, FExecutor.SetOptions);
+  ValueLabelOutput    := ValueLabelTypeFromOptionList(Command.Options, FExecutor.SetOptions);
+  sColor              := ChartColorsFromOptions(Command.Options, FExecutor.SetOptions);
+  VarNames            := Command.VariableList.GetIdentsAsList;
+  DFVars              := Command.VariableList.GetIdentsAsList; // variable order may change when adding weight var
+  StratVariable       := TStringList.Create;
+  cOptions            := TOptionList.Create;
```

Comment:
If I were to be pedantic I would say this should have been fixed in the TOptionList. Like implementing a method that copied this structure to a new optionlist.
But this is fine and another solution can be implemented at a later stage if we care.

Reply:
I agree. I'll create a new branch and add an assign method to TOptionList, which seems to be the standard way of doing this. I might do the same for VariableList, because I have to create a second variable list to add variable 2 of the bar charts as a stratifying variable. Again, don't want to alter Command.VariableList.